apps/framework-cli/src/cli/routines/openapi.rs

```diff
@@ -0,0 +1,231 @@
+use crate::framework::core::infrastructure::api_endpoint::APIType;
+use crate::framework::core::infrastructure::table::ColumnType;
+use crate::framework::core::infrastructure_map::InfrastructureMap;
+use crate::infrastructure::olap::clickhouse_alt_client::{get_pool, retrieve_infrastructure_map};
+use crate::project::Project;
+use crate::utilities::constants::OPENAPI_FILE;
+
+use chrono::Local;
+use serde::{Deserialize, Serialize};
+use serde_yaml;
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Write;
+use std::sync::Arc;
+
+#[derive(Serialize, Deserialize)]
+struct OpenAPI {
+    openapi: String,
+    info: Info,
+    servers: Vec<Server>,
+    paths: HashMap<String, PathItem>,
+    components: Components,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Info {
+    title: String,
+    version: String,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Server {
+    url: String,
+    description: Option<String>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct PathItem {
+    #[serde(rename = "post")]
+    post: Option<Operation>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Operation {
+    summary: String,
+    #[serde(rename = "requestBody")]
+    request_body: Option<RequestBody>,
+    responses: HashMap<String, Response>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct RequestBody {
+    content: HashMap<String, MediaType>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct MediaType {
+    schema: Schema,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+struct Schema {
+    #[serde(rename = "type")]
+    schema_type: String,
+    properties: HashMap<String, Property>,
+    required: Vec<String>,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+struct Property {
+    #[serde(rename = "type")]
+    property_type: String,
+    example: Option<serde_json::Value>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Response {
+    description: String,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Components {
+    schemas: HashMap<String, Schema>,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum OpenAPIError {
+    #[error("Error connecting to storage")]
+    Connection,
+    #[error("Error retrieving current state")]
+    Retrieval,
+    #[error("No state found")]
+    NoState,
+    #[error("Failed to save OpenAPI spec to file: {0}")]
+    Save(String),
+}
+
+pub async fn openapi(project: &Arc<Project>) -> Result<(), OpenAPIError> {
+    let mut client = get_pool(&project.clickhouse_config)
+        .get_handle()
+        .await
+        .map_err(|_| OpenAPIError::Connection)?;
+
+    let infra = retrieve_infrastructure_map(&mut client, &project.clickhouse_config)
```

Comment:
Do we want to build more on top of the clickhouse map or should we load it on top of the redis content?

Reply:
I may have missed the convo on why infra map is in both ch & redis. Could you or @phiSgr give me some context?

Reply:
the infra map stored on redis was added to allow boreal to read it.

Reply:
Ah okay. I don't see `infrastructure_map` in boreal yet, is that feature still on the docket? @okane16 , would you happen to know?

I don't have a preference where we get the infra map from for local dev. If it's redis, I think there's some additional work to deserialize the protobuf.

Reply:
@phiSgr don't we have the client already setup?